// kayn/dist/lib/Endpoints/MatchEndpoint/MatchlistEndpoint.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _MatchSuperclass = require('./MatchSuperclass');

var _MatchSuperclass2 = _interopRequireDefault(_MatchSuperclass);

var _Request = require('../../RequestClient/Request');

var _Request2 = _interopRequireDefault(_Request);

var _methodNames = require('../../Enums/method-names');

var _methodNames2 = _interopRequireDefault(_methodNames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class MatchlistEndpoint extends _MatchSuperclass2.default {
    constructor(config, limiter) {
        super();

        this.config = config;

        this.by = {
            accountID: this.accountID.bind(this)
        };

        this.Recent = {
            by: {
                accountID: this.getRecentMatchlistByAccountID.bind(this)
            }
        };

        this.limiter = limiter;
    }

    /**
    * Get matchlist for games played on given account ID and platform ID and filtered using given filter parameters, if any.
    * 
    * Implements /lol/match/v3/matchlists/by-account/{accountId}.
    *  
    * @param {number} accountID - The account ID of the summoner. 
    */
    accountID(accountID) {
        return new _Request2.default(this.config, this.serviceName, `matchlists/by-account/${accountID}`, _methodNames2.default.MATCH.GET_MATCHLIST, 'GET', this.limiter);
    }

    /**
    * Get matchlist for last 20 matches played on given account ID and platform ID.  
    * 
    * Implements /lol/match/v3/matchlists/by-account/{accountId}/recent.
    *  
    * @param {number} accountID - The account ID of the summoner. 
    */
    getRecentMatchlistByAccountID(accountID) {
        return new _Request2.default(this.config, this.serviceName, `matchlists/by-account/${accountID}/recent`, _methodNames2.default.MATCH.GET_RECENT_MATCHLIST, 'GET', this.limiter);
    }
}

exports.default = MatchlistEndpoint;
