// kayn/dist/lib/KaynConfig.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.KAYN_CONFIG_STRUCT = exports.DEFAULT_KAYN_CONFIG = undefined;

var _superstruct = require('superstruct');

var _methodNames = require('./Enums/method-names');

var _methodNames2 = _interopRequireDefault(_methodNames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_KAYN_CONFIG = exports.DEFAULT_KAYN_CONFIG = {
    region: 'na',
    debugOptions: {
        isEnabled: true,
        showKey: false,
        loggers: {}
    },
    requestOptions: {
        shouldRetry: true,
        numberOfRetriesBeforeAbort: 3,
        delayBeforeRetry: 1000,
        burst: false
    },
    cacheOptions: {
        cache: null,
        ttls: {}
    }
};

const ttlsValidator = {};
Object.keys(_methodNames2.default).forEach(s => {
    Object.keys(_methodNames2.default[s]).forEach(t => ttlsValidator[_methodNames2.default[s][t]] = 'number?');
});

const KAYN_CONFIG_STRUCT = exports.KAYN_CONFIG_STRUCT = (0, _superstruct.struct)({
    region: 'string',
    debugOptions: {
        isEnabled: 'boolean',
        showKey: 'boolean',
        loggers: 'any'
    },
    requestOptions: {
        shouldRetry: 'boolean',
        numberOfRetriesBeforeAbort: 'number',
        delayBeforeRetry: 'number',
        burst: 'boolean'
    },
    cacheOptions: {
        cache: 'any',
        ttls: ttlsValidator
    },
    key: 'string'
});
